core/tests/ledger_cleanup.rs

```diff
@@ -338,27 +338,45 @@ mod tests {
                     let mut num_shreds = 0;
                     let mut max_speed = 0f32;
                     let mut min_speed = f32::MAX;
+                    let (mut new_shreds, _) = make_many_slot_shreds(
+                        1, batch_size_slots, shreds_per_slot);
                     loop {
-                        let batch_id = shared_batch_id.fetch_add(1, Ordering::SeqCst);
+                        let batch_id = shared_batch_id.fetch_add(1, Ordering::Relaxed);
                         let start_slot = batch_id * batch_size_slots;
-                        let len = batch_id;
                         if start_slot >= benchmark_slots {
                             break;
                         }
+                        let len = batch_id;
 
-                        let new_shreds = if pre_generate_data {
+                        let br = if pre_generate_data {
                             let mut sl = cloned_shreds.lock().unwrap();
                             if let Some(shreds_from_queue) = sl.pop_front() {
-                                shreds_from_queue
+                                cloned_blockstore.insert_shreds(
+                                    shreds_from_queue, None, false).unwrap()
                             } else {
+                                // If the queue is empty, we're done!
                                 break;
                             }
                         } else {
-                            let (generated_shreds, _) = make_many_slot_shreds(
-                                start_slot, batch_size_slots, shreds_per_slot);
-                            generated_shreds
+                            let mut slot_id = start_slot;
+                            for shred in new_shreds.iter_mut() {
+                                // Whenever we sees a shared with index 0,
+                                // it means it is a new slot.
+                                if shred.index() == 0 {
+                                    slot_id += 1;
+                                }
+                                shred.set_slot(slot_id);
+                            }
```

Comment:
Having `slot_id` incremented prior to setting slot in the shreds means that we'll miss slot `start_slot`.

For example, let's suppose we get `batch_id == 0` such that `start_slot == 0` as well. We'll start with `slot_id == start_slot == 0`, and on the first iteration of this loop, we'll enter the if statement and bump `slot_id` immediately. So, we'll actually produce slots `[batch_id + 1, .... , batch_id + batch_size_slots]` instead of `[batch_id, ... , batch_id + batch_size_slots - 1]`.

Possible alternative:
```
let mut new_slot = start_slot;
let mut num_updated_shreds = 0;
for shred in new_shreds.iter_mut() {
  shred.set_slot(new_slot);
  num_updated_shreds += 1;
  if num_updated_shreds % shreds_per_slot == 0 {
    new_slot += 1;
  }
}
```

I don't think this would affect anything with actually inserting shreds (each writer thread would produce shreds with the +1 offset so they shouldn't be duplicates), but possibly cause us an off-by-one error in the ledger cleaning up shreds?

Reply:
I think starting from slot 1 should be fine, but slot 1 will mark slot 0 as its parent, which does not exist.

One workaround might be pre-inserting slot 0 before the benchmark so that we can have slot 0 without a parent, and having the remaining slots with -1 slot as their parent.  I can add pre-insert one slot, but I think overall it won't affect the benchmark.

The cleanup service should be fine, as it only compares whether we should purge by checking [`root - last_purge_slot <= purge_interval`. ](https://github.com/yhchiang-sol/solana/blob/bb97c8fdcd3ec43979f25c18c49587c91089723d/core/src/ledger_cleanup_service.rs#L181)

Reply:
> One workaround might be pre-inserting slot 0 before the benchmark so that we can have slot 0 without a parent, and having the remaining slots with -1 slot as their parent. I can add pre-insert one slot, but I think overall it won't affect the benchmark.

I'd argue we shouldn't do this to keep pre-generate vs. no-pre-generate cases as similar as possible (no-pre-generate would have an extra slot); the ledger should have same set of shreds after the test regardless of pre-generate value. Even though this extra slot would be inserted outside of the timed section, it could hypothetically have some indirect effect (pushing us over a boundary for having to flush an extra memtable) in the timed section. While this is unlikely, we can make it a 0% chance by making the change

> The cleanup service should be fine, as it only compares whether we should purge by checking root - last_purge_slot <= purge_interval.

Agreed that it is fine in its current state; but, if that logic changed to be more "exact" instead of using an interval, then this issue could become a bug. Additionally, this line of code
https://github.com/solana-labs/solana/blob/ce6c76e45feffe49096eed0c829d6d8b747cdaba/core/tests/ledger_cleanup.rs#L412

assumes that we do not miss `start_slot` so another spot where we could run into trouble if things change and an "off-by-1" could hurt us more.

Lastly, I think getting rid of the offset makes the code more "predictable". People might look at / run tests to try to understand how the codebase works, and skipping `start_slot` could be a minor point of confusion that doesn't need to be

Reply:
The pain-in-the-ass is that the shred in slot-0 is slightly different from shreds in other slots, as shreds in slot 0 do not have a parent slot, while others have slot - 1 as their parent slot.

https://github.com/yhchiang-sol/solana/blob/b18462737e660c2792899d9234f734ee64250112/ledger/src/blockstore.rs#L4050

One way to really fix it without any side effect is to generate two shred templates:  one for the first slot, and one for the remaining slots.  This should make the two settings equivalent.

Reply:
>The pain-in-the-ass is that the shred in slot-0 is slightly different from shreds in other slots, as shreds in slot 0 do not have a parent slot, while others have slot - 1 as their parent slot.

Good point - to sum up what we discussed in DM, thanks for reminding me that the `parent_offset` flag is what is actually encoded in the shred. Using the test function to generate shreds, this value is `0` for slot `0` and the previous slot for all other slots. Because we do want `parent_flag` to be set correctly in our inserted shreds, our "template" batch needs to start at a non-zero slot. You mentioned that your next update will fix this, so I'll mark this as resolved when you push your next update